holoviews/operation/datashader.py

```diff
@@ -1725,3 +1725,193 @@ class directly_connect_edges(_connect_edges, connect_edges):
+            vdims = self._vdims(raster, df)
+            return Points(df, kdims=raster.kdims, vdims=vdims)
+
+        masked = self._mask_dataframe(raster, x, y, xdelta, ydelta)
+        dist_sorted = self._sort_by_distance(raster, masked, x, y)
+        df = self.p.points_transformer(self.p.transformer(dist_sorted))
+        self.hits = self.p.hits_transformer(self.p.transformer(dist_sorted))
+        vdims = self._vdims(raster, df)
+        return Points(df, kdims=raster.kdims, vdims=vdims).iloc[:self.p.point_count]
+
+    @property
+    def mask(self):
+        return inspect_mask.instance(pixels=self.p.pixels)
+
+    @classmethod
+    def _vdims(cls, raster, df):
+        ds = raster.dataset
+        if 'spatialpandas' in ds.interface.datatype:
+            coords = [ds.interface.geo_column(ds.data)]
+        else:
+            coords = [kd.name for kd in raster.kdims]
+        return [col for col in df.columns if col not in coords]
+
+    @classmethod
+    def _empty_df(cls, dataset):
+        if 'dask' in dataset.interface.datatype:
+            return dataset.data._meta.iloc[:0]
+        elif dataset.interface.datatype in ['pandas', 'geopandas', 'spatialpandas']:
+            return dataset.data.head(0)
+        return dataset.iloc[:0].dframe()
+
+    @classmethod
+    def _distance_args(cls, element, x_range, y_range,  pixels):
+        ycount, xcount =  element.interface.shape(element, gridded=True)
+        x_delta = abs(x_range[1] - x_range[0]) / xcount
+        y_delta = abs(y_range[1] - y_range[0]) / ycount
+        return (x_delta*pixels, y_delta*pixels)
+
+    @classmethod
+    def _mask_dataframe(cls, raster, x, y, xdelta, ydelta):
+        "Mask the dataframe with size around x and y"
+        ds = raster.dataset
+        x0, x1, y0, y1 = x-xdelta, x+xdelta, y-ydelta, y+ydelta
+        if 'spatialpandas' in ds.interface.datatype:
+            df = ds.data.cx[x0:x1, y0:y1]
+            return df.compute() if hasattr(df, 'compute') else df
+        xdim, ydim = raster.kdims
+        query = {xdim.name: (x0, x1), ydim.name: (y0, y1)}
+        return ds.select(**query).dframe()
+
+    @classmethod
+    def _get_xs_ys(cls, raster, df):
+        "Returns the x and y arrays or series for the given dataframe"
+        ds = raster.dataset
+        if 'spatialpandas' in ds.interface.datatype:
+            geom = ds.interface.geo_column(ds.data)
+            return df[geom].array.x, df[geom].array.y
+        xdim, ydim = raster.kdims
+        return df[xdim].values, df[ydim].values
+
+    @classmethod
+    def _sort_by_distance(cls, raster, df, x, y):
+        """
+        Returns a dataframe of hits within a given mask around a given
+        spatial location, sorted by distance from that location.
+        """
+        xs, ys = cls._get_xs_ys(raster, df)
+        dx, dy = xs - x, ys - y
+        distances = pd.Series(dx*dx + dy*dy)
+        return df.iloc[distances.argsort().values]
```

Comment:
I don't know how expensive it is to calculate distances like this, but if it's quicker to run the query multiple times (1x1, 3x3, etc. up to max_points) and abort early, we should! In the worst case (clicking in a dense region), which is presumably where we have to worry the most about performance, I'd expect the 1x1 region nearly always to have at least one datapoint, so the other cases would only be reached in sparser regions where I'd think it's cheaper to do the query anyway.

Reply:
Shouldn't we just filter by max points before computing distances? We've already established that they are in the target region.

Reply:
Ah sorry then you can't establish which is closest. Ignore me.

Reply:
> I don't know how expensive it is to calculate distances like this...

It will be O(N) in the number of hits found. If you have a million hits I imagine it will take a while to compute but that seems unlikely.

> ... but if it's quicker to run the query multiple times (1x1, 3x3, etc. up to max_points) and abort early, we should

Honestly, I think this is tricky to reason about because the performance is going to depend a lot on the distribution of the data being inspected. 

> 1x1, 3x3, etc. up to max_points

I'm not sure I follow. Those look like mask sizes to me so I don't see how that relates to `max_points`. Unless you meant up to the full mask size? If so, it sounds like you mean to query a single pixel mask, and if there are insufficient hits, try 2x2 etc. 

Such an algorithm would need multiple `.cx` queries which is likely to me *much* more expensive than computing the distances unless there are a huge number of matches in the mask. The possibly optimization I had in mind was different: do a chunk of distances and if you have `max_points` rows that are found within the center pixel, return them (basically this says any samples within the center pixel are considered equivalently good no matter where the cursor is in that pixel). Note that if you consider a raster to be an approximation to an underlying density, finding the minimum distance (as it does now) is the most correct thing to do...

I agree there may be cases where the distance computation becomes the limiting factor but my intuition is that the `.cx` filter will be the far bigger limiting factor in any realistic dataset (yes, you could try to inspect a dataframe with 100 million points with exactly the same position but this doesn't seem very realistic).

Reply:
Agree that this sounds significantly **more** expensive rather than less, particularly in the worst case. I don't see **any** scenario where repeated spatial queries will be faster than computing distances.

Reply:
I must be missing something here. The worst case is when you click on a dense region of the plot, right? I.e. somewhere there are millions of points per pixel?  Is so how could looking in a 1x1 pixel area first be bad?  With a very high probability, there will be a point in the 1x1 pixel area and it will then get returned, leaving the total search space lower than if you had done a smaller mask.  It's true that in some other cases one will have to do multiple queries, but aren't those the easy cases, where there are few points?

Reply:
>It's true that in some other cases one will have to do multiple queries, but aren't those the easy cases, where there are few points?

That's assuming a lot about the data distribution no? In any case though I think you aren't appreciating the difference in cost between the two operations, the 1x1 query will take pretty much the same amount of time as the 3x3 query. So yes, you'll save on the distance calculation but that step is (almost) never going to be expensive when compared to the spatial query, e.g. it takes 70 ms on 10 million points. So you're trading a small bit of speedup in the distance calculation in the best case against an `~N x` slowdown in the worst case, where N is the size of the `N pixel x N pixel` mask.

Reply:
> Is so how could looking in a 1x1 pixel area first be bad?

There is no ordering in which you look at pixels as it all works through a `cx` slice. Looking at one pixel takes just as long as a 100 (assuming you are in the same partition): you need to pass through all the rows and filter out the ones outside the chosen region.